### PR TITLE
[ros2] Add Stage property to assign a list of controllers to use when executing the planned trajectory

### PR DESCRIPTION
**Pull request by schornakj**
This is essentially a ROS2 version of the work in #123, plus some additions based on @henningkayser's review of that PR.

- Add a new `TrajectoryExecutionInfo` ROS message to hold info used when executing the trajectory. Right now it just contains a vector of strings to define controller names used with the trajectory. Add an instance of this message as a field in the `SubTrajectory` message.
- Add a new `TrajectoryExecutionInfo` struct to hold this info as an MTC property.
- Propagate the `"trajectory_execution_info"` property from each stage into the `SubTrajectory` messages output when the task is planned.
- Modify the `ExecuteTaskSolution` capability to copy the controller names into the `RobotTrajectory` prior to execution.

### CODE REVIEW

**Comment by IgorUnlimited** [comment]
Pick/place still not work. Planning failed ... after running:

1) `ros2 launch moveit_task_constructor_demo demo.launch.py`
2) `ros2 launch moveit_task_constructor_demo pickplace.launch.py`

**Comment by schornakj** [comment]
> Pick/place still not work. Planning failed ... after running:
> 
>     1. `ros2 launch moveit_task_constructor_demo demo.launch.py`
> 
>     2. `ros2 launch moveit_task_constructor_demo pickplace.launch.py`

Was your comment meant for #350? This PR wasn't intended as a fix for the issue described over there.

**Comment by schornakj** [comment]
I'll tag @JafarAbdi here since we were talking about this PR earlier today.

**Comment by sjahr** [comment]
@rhaschke @v4hn Other than the conflict, what is blocking this PR from being merged? I saw this comment in #26:
> Anyway, I was not happy with that approach because it focused only on controller names.

What else would you add? To move forward I think it would be good to review and merge this PR to make this usable and build on top of it. A nice follow-up PR could involve hiding the actual controller names behind more abstracted config names such as complaint trajectory execution (e.g. JTC + admittance control) and offer such an API in MoveIt itself

File: core/include/moveit/task_constructor/stage.h
Comment (sjahr) [inline core/include/moveit/task_constructor/stage.h:205]: Nit: Document functions individually

File: core/include/moveit/task_constructor/trajectory_execution_info.h
Comment (sjahr) [inline core/include/moveit/task_constructor/trajectory_execution_info.h:42]: Is this used anywhere?

File: core/include/moveit/task_constructor/trajectory_execution_info.h
Comment (sjahr) [inline core/include/moveit/task_constructor/trajectory_execution_info.h:52]: Can you add documentation for this class and its members

**Review by sjahr — COMMENTED** [review]
This is useful to pass information downwards, how a trajectory should be used. I'd be in favor of merging it once more documentation is added

File: core/src/container.cpp
Comment (v4hn) [inline core/src/container.cpp:286]: This would overwrite whatever was specified for the stage, right?

File: core/src/storage.cpp
Comment (v4hn) [inline core/src/storage.cpp:195]: This way the execution info cannot be set *per solution*, but it is entirely valid that the same stage would generate solutions with, e.g., a left or a right arm. Accordingly, the controllers that you would need to execute the solution varies per solution.

**Review by v4hn — COMMENTED** [review]
I still believe this is a special case of solution properties which [Robert summarized recently](https://github.com/ubi-agni/moveit_task_constructor/issues/26) and I think it doesn't make sense to merge partial support for this very specific use or modifying the messages for it.
For the execution through `MoveItCpp` there's not much missing to get the generic properties to work.

File: core/src/container.cpp
Comment (sjahr) [inline core/src/container.cpp:286]:
```suggestion
	if (stage->trajectoryExecutionInfo().controller_names.empty()) {
		stage->setTrajectoryExecutionInfo(this->trajectoryExecutionInfo());
	}
```

File: capabilities/src/execute_task_solution_capability.cpp
Comment (sjahr) [inline capabilities/src/execute_task_solution_capability.cpp:173]:
```suggestion
		exec_traj.trajectory = std::make_shared<robot_trajectory::RobotTrajectory>(model, group);
		exec_traj.trajectory->setRobotTrajectoryMsg(state, sub_traj.trajectory);
		exec_traj.controller_name = sub_traj.execution_info.controller_names;
```

File: core/include/moveit/task_constructor/trajectory_execution_info.h
Comment (sjahr) [inline core/include/moveit/task_constructor/trajectory_execution_info.h:42]:
```suggestion
```

File: core/src/storage.cpp
Comment (sjahr) [reply core/src/storage.cpp:234]: Is there a solution available to pick the trajectory info per solution/ any need for it at the moment? In order to unblock this PR I think this could be addressed in a follow-up issue when there is the need to specify trajectory info per solution.

**Review by sjahr — APPROVED** [review]
@v4hn @rhaschke I think we should merge this. We've been successfully using these changes for ~1 year in our MTC fork. This PR might only cover a special case of solution properties but nobody has started working on #26 so far and merging this won't prevent anyone from refactoring and generalizing this.

File: core/src/storage.cpp
Comment (rhaschke) [reply core/src/storage.cpp:234]: We would need to register custom user callbacks to fill those fields in a dynamic fashion.

**Review by rhaschke — COMMENTED** [review]
Generally, I'm fine to merge this temporarily. But be prepared that we change/generalize the API in future. I have to admit that MTC wasn't very high on my prior list lately.

While you rely on #123, I'm surprised to not see a cherry-pick of these commits.
To honor @llach original work on this (and keep his authorship), I suggest to cherry-pick (and squash) #123 and then add your modificiations on top of it. Please also rebase to the ros2 branch, such that we can merge via a true merge-commit.

**Comment by sjahr** [comment]
@rhaschke Done :heavy_check_mark:

File: core/include/moveit/task_constructor/trajectory_execution_info.h
Comment (rhaschke) [inline core/include/moveit/task_constructor/trajectory_execution_info.h:51]: Why don't you simply define a synonym:
`using TrajectoryExecutionInfo = moveit_task_constructor_msgs::msg::TrajectoryExecutionInfo`?

File: core/src/container.cpp
Comment (rhaschke) [inline core/src/container.cpp:370]: This is not the correct way of inheriting properties from parent to child (I think this is intended here).
Users should make this inheritance explicit _in their code_ via:
`Property::configureInitFrom(Stage::PARENT, "trajectory_execution_info");`

File: capabilities/src/execute_task_solution_capability.cpp
Comment (rhaschke) [inline capabilities/src/execute_task_solution_capability.cpp]: The names (`_`) are changed back and forth. Please further cleanup history to have minimal changes per commit. 
This also facilitates reviewing :wink:

**Review by rhaschke — COMMENTED** [review]
Thanks for cleaning up the history. Please squash the `_` fixups into the first commit.
I think the 3rd commit should be dropped completely (see comment below).

File: core/src/container.cpp
Comment (sjahr) [reply core/src/container.cpp:370]: The intend here is to make inheriting TrajectoryExecutionInfo from the parent in the container default except, the stage has explicitly configured it's own trajectory execution info. If a user has already done this via configureInitFrom, nothing happens here. Do you think that is a bad default?

**Comment by sjahr** [comment]
@rhaschke Can you give this a final review?

**Review by sea-bass — APPROVED** [review]
This has been working for us for a while, so I'm inclined to approve it so we can move away from our fork and use the upstream MTC.

@rhaschke if there is some concrete changes in this PR you would like to see, please let @sjahr know before he merges it... or if you prefer to revisit this later, that works too.

File: core/src/container.cpp
Comment (rhaschke) [reply core/src/container.cpp:370]: Yes, I do consider this a bad default, because it contradicts the philosophy of _explicit_ property inheritance, which @v4hn requested from the very beginning. We should not deviate from that philosophy. Please revert that part.

Comment (sjahr) [reply core/src/container.cpp:370]: https://github.com/ros-planning/moveit_task_constructor/pull/502

**Comment by rhaschke** [comment]
Looks like this PR introduced a segfault in unit tests:
https://github.com/ros-planning/moveit_task_constructor/actions/runs/6730632997

Please investigate!

**Comment by sjahr** [comment]
:thinking: Why wasn't it caught by CI?

**Comment by rhaschke** [comment]
Not sure. Maybe it is just due to a recent change in MoveIt2?

**Comment by sjahr** [comment]
I'll investigate it :+1:

**Pull request merged**
